File: W11/class-problems/pt_in_poly.cpp
#include <iostream>
#include <vector>
using namespace std;

#define ll long long

pair<ll,ll> operator-(const pair<ll,ll> &p1, const pair<ll,ll> &p2) {
    return {p1.first - p2.first, p1.second - p2.second};
}

ll det(const pair<ll,ll> &p1, const pair<ll,ll> &p2) {
    return p1.first * p2.second - p1.second*p2.first;
}

ll norm_sq(const pair<ll,ll> &p) {
    return p.first*p.first + p.second*p.second;
}

int main() {
    int n,m;
    while((cin >> n) && n) {
        vector<pair<ll,ll>> poly;
        for(int i = 0; i < n; i++) {
            ll x,y;
            cin >> x >> y;
            poly.push_back({x,y});
        }
        poly.push_back(poly[0]);

        cin >> m;
        for(int z = 0; z < m; z++) {
            ll x,y;
            cin >> x >> y;
            int count = 0;
            bool on = false;
            for(int i = 0; i < n; i++) {
                pair<ll,ll> t,b;
                if((poly[i].second > poly[i+1].second) ||
                    (poly[i].second == poly[i+1].second && poly[i].first < poly[i+1].first)) {
                    t = poly[i];
                    b = poly[i+1];
                } else {
                    t = poly[i+1];
                    b = poly[i];
                }
                if(y < b.second || y > t.second) continue;
                if(b.second == t.second) {
                    if(x >= t.first && x <= b.first) {
                        on = true;
                        break;
                    }
                } else if (y == t.second) {
                    if(x == t.first) {
                        on = true;
                        break;
                    }
                } else if (y == b.second) {
                    if(x == b.first) {
                        on = true;
                        break;
                    } else if (x < b.first) {
                        count++;
                    }
                } else { // general case
                    pair<ll,ll> p({x,y});
                    ll cp = det(b-t,p-t);
                    if(cp == 0) {
                        on = true;
                        break;
                    } else if (cp < 0) {
                        count++;
                    }
                }
            }
            if(on) {
                cout << "on" << endl;
            } else if (count%2 == 1) {
                cout << "in" << endl;
            } else {
                cout << "out" << endl;
            }
        }
    }
    return 0;
}

File: W11/class-problems/pt_in_poly.py
"""

    https://open.kattis.com/problems/pointinpolygon

"""

def subtract(p1, p2):
    return (p1[0] - p2[0], p1[1] - p2[1])

def det(p1, p2):
    return p1[0] * p2[1] - p1[1] * p2[0]

def norm_sq(p):
    return p[0] * p[0] + p[1] * p[1]

def main():
    while True:
        line = input().strip()
        if not line:
            break
        n = int(line)
        if n == 0:
            break

        poly = []
        for _ in range(n):
            x, y = map(int, input().split())
            poly.append((x, y))

        # close polygon
        poly.append(poly[0])

        m = int(input())
        for _ in range(m):
            x, y = map(int, input().split())
            count = 0
            on = False

            for i in range(n):
                # assign top/bottom (t = top point, b = bottom point)
                if (poly[i][1] > poly[i+1][1]) or \
                   (poly[i][1] == poly[i+1][1] and poly[i][0] < poly[i+1][0]):
                    t = poly[i]
                    b = poly[i+1]
                else:
                    t = poly[i+1]
                    b = poly[i]

                # point y outside vertical span
                if y < b[1] or y > t[1]:
                    continue

                # horizontal edge
                if b[1] == t[1]:
                    if x >= min(t[0], b[0]) and x <= max(t[0], b[0]):
                        on = True
                        break

                # check if point is exactly on top point
                elif y == t[1]:
                    if x == t[0]:
                        on = True
                        break

                # check if point is exactly on bottom point
                elif y == b[1]:
                    if x == b[0]:
                        on = True
                        break
                    elif x < b[0]:
                        count += 1

                else:
                    # general case ray intersection test
                    p = (x, y)
                    cp = det(subtract(b, t), subtract(p, t))
                    if cp == 0:
                        on = True
                        break
                    elif cp < 0:
                        count += 1

            if on:
                print("on")
            elif count % 2 == 1:
                print("in")
            else:
                print("out")

if __name__ == "__main__":
    main()